// live-set/filter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = filter;

var _ = _interopRequireDefault(require("."));

/* eslint-disable no-unused-vars, no-redeclare */

/*:: export default filter; */

/*:: ` */
function filter(liveSet, cb) {
  return new _.default({
    scheduler: liveSet.getScheduler(),
    read: function read() {
      var ret = new Set();
      liveSet.values().forEach(function (value) {
        if (cb(value)) {
          ret.add(value);
        }
      });
      return ret;
    },
    listen: function listen(setValues, controller) {
      var passedFilter = new Set();
      var sub = liveSet.subscribe({
        start: function start() {
          var initialValues = new Set();
          liveSet.values().forEach(function (value) {
            if (cb(value)) {
              passedFilter.add(value);
              initialValues.add(value);
            }
          });
          setValues(initialValues);
        },
        next: function next(changes) {
          changes.forEach(function (change) {
            if (change.type === 'add') {
              if (cb(change.value)) {
                passedFilter.add(change.value);
                controller.add(change.value);
              }
            } else if (change.type === 'remove') {
              if (passedFilter.has(change.value)) {
                passedFilter.delete(change.value);
                controller.remove(change.value);
              }
            }
          });
        },
        error: function error(err) {
          controller.error(err);
        },
        complete: function complete() {
          controller.end();
        }
      });
      return sub;
    }
  });
}
/*:: ` */


module.exports = exports.default;
module.exports.default = exports.default;
