// live-set/map.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = map;

var _ = _interopRequireDefault(require("."));

function map(liveSet, cb) {
  return new _.default({
    scheduler: liveSet.getScheduler(),
    read: function read() {
      var s = new Set();
      liveSet.values().forEach(function (value) {
        s.add(cb(value));
      });
      return s;
    },
    listen: function listen(setValues, controller) {
      var m = new Map();
      var sub = liveSet.subscribe({
        start: function start() {
          var s = new Set();
          liveSet.values().forEach(function (value) {
            var newValue = cb(value);
            m.set(value, newValue);
            s.add(newValue);
          });
          setValues(s);
        },
        next: function next(changes) {
          changes.forEach(function (change) {
            if (change.type === 'add') {
              var newValue = cb(change.value);
              m.set(change.value, newValue);
              controller.add(newValue);
            } else if (change.type === 'remove') {
              var _newValue = m.get(change.value);

              if (!_newValue) throw new Error('removed item not in liveset');
              m.delete(change.value);
              controller.remove(_newValue);
            }
          });
        },
        error: function error(err) {
          controller.error(err);
        },
        complete: function complete() {
          controller.end();
        }
      });
      return sub;
    }
  });
}

module.exports = exports.default;
module.exports.default = exports.default;
