// live-set/mapWithRemoval.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = mapWithRemoval;

var _ = _interopRequireDefault(require("."));

function mapWithRemoval(input, cb) {
  var output = new _.default({
    scheduler: input.getScheduler(),
    read: function read() {
      throw new Error('mapWithRemoval liveset may not be read while inactive');
    },
    listen: function listen(setValues, controller) {
      var m = new Map();
      var sub = input.subscribe({
        start: function start() {
          var s = new Set();
          input.values().forEach(function (value) {
            var resolve;
            var removal = new Promise(function (_resolve) {
              resolve = _resolve;
            });
            /*:: if (!resolve) throw new Error(); */

            var newValue = cb(value, removal);
            m.set(value, {
              newValue: newValue,
              resolve: resolve
            });
            s.add(newValue);
          });
          setValues(s);
        },
        next: function next(changes) {
          changes.forEach(function (change) {
            if (change.type === 'add') {
              var resolve;

              var _removal = new Promise(function (_resolve) {
                resolve = _resolve;
              });
              /*:: if (!resolve) throw new Error(); */


              var newValue = cb(change.value, _removal);
              m.set(change.value, {
                newValue: newValue,
                resolve: resolve
              });
              controller.add(newValue);
            } else if (change.type === 'remove') {
              var entry = m.get(change.value);
              if (!entry) throw new Error('removed item not in liveset');
              var _newValue = entry.newValue,
                  _resolve2 = entry.resolve;

              _resolve2();

              m.delete(change.value);
              controller.remove(_newValue);
            }
          });
        },
        error: function error(err) {
          controller.error(err);
        },
        complete: function complete() {
          controller.end();
        }
      });
      return {
        unsubscribe: function unsubscribe() {
          sub.unsubscribe();
          m.forEach(function (_ref) {
            var resolve = _ref.resolve;
            resolve();
          });
        },
        pullChanges: function pullChanges() {
          sub.pullChanges();
        }
      };
    }
  });
  return output;
}

module.exports = exports.default;
module.exports.default = exports.default;
