// live-set/merge.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = merge;

var _ = _interopRequireDefault(require("."));

function merge(liveSets) {
  return new _.default({
    scheduler: liveSets[0] ? liveSets[0].getScheduler() : undefined,
    read: function read() {
      var s = new Set();
      liveSets.forEach(function (liveSet) {
        liveSet.values().forEach(function (value) {
          s.add(value);
        });
      });
      return s;
    },
    listen: function listen(setValues, controller) {
      var initialValues = new Set();
      var subs = new Set();
      var doneSubscribing = false;
      liveSets.forEach(function (liveSet) {
        var sub;
        liveSet.subscribe({
          start: function start(_sub) {
            sub = _sub;
            subs.add(sub);
            liveSet.values().forEach(function (value) {
              initialValues.add(value);
            });
          },
          next: function next(changes) {
            changes.forEach(function (change) {
              if (change.type === 'add') {
                controller.add(change.value);
              } else if (change.type === 'remove') {
                controller.remove(change.value);
              }
            });
          },
          error: function error(err) {
            controller.error(err);
          },
          complete: function complete() {
            subs.delete(sub);

            if (doneSubscribing && subs.size === 0) {
              controller.end();
            }
          }
        });
      });
      setValues(initialValues);
      doneSubscribing = true;

      if (subs.size === 0) {
        controller.end();
      }

      return {
        unsubscribe: function unsubscribe() {
          subs.forEach(function (sub) {
            sub.unsubscribe();
          });
        },
        pullChanges: function pullChanges() {
          subs.forEach(function (sub) {
            sub.pullChanges();
          });
        }
      };
    }
  });
}

module.exports = exports.default;
module.exports.default = exports.default;
