// live-set/toValueObservable.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = toValueObservable;

var _ = _interopRequireDefault(require("."));

var _zenObservable = _interopRequireDefault(require("zen-observable"));

function toValueObservable(liveSet) {
  return new _zenObservable.default(function (observer) {
    var resolvers = new Map();

    function addedItem(value) {
      var resolve;
      var removal = new Promise(function (_resolve) {
        resolve = _resolve;
      });
      resolvers.set(value, resolve);
      var valueWithRemoval = {
        value: value,
        removal: removal
      };
      observer.next(valueWithRemoval);
    }

    function removedItem(value) {
      var resolver = resolvers.get(value);
      if (!resolver) throw new Error('Resolver not found in map for value');
      resolvers.delete(value);
      resolver();
    }

    var sub = liveSet.subscribe({
      start: function start(sub) {
        var _iteratorNormalCompletion = true;
        var _didIteratorError = false;
        var _iteratorError = undefined;

        try {
          for (var _iterator = liveSet.values()[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
            var value = _step.value;
            if (sub.closed) break;
            addedItem(value);
          }
        } catch (err) {
          _didIteratorError = true;
          _iteratorError = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion && _iterator.return != null) {
              _iterator.return();
            }
          } finally {
            if (_didIteratorError) {
              throw _iteratorError;
            }
          }
        }
      },
      next: function next(changes) {
        changes.forEach(function (change) {
          if (change.type === 'add') {
            addedItem(change.value);
          } else if (change.type === 'remove') {
            removedItem(change.value);
          }
        });
      },
      error: function error(err) {
        observer.error(err);
      },
      complete: function complete() {
        observer.complete();
      }
    });
    return function () {
      sub.unsubscribe();
      resolvers.forEach(function (resolver) {
        resolver();
      });
    };
  });
}

module.exports = exports.default;
module.exports.default = exports.default;
