// tag-tree/js/TagTreeNode.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _liveSet = _interopRequireDefault(require("live-set"));

var TagTreeNode =
/*#__PURE__*/
function () {
  function TagTreeNode(init) {
    var _this = this;

    (0, _classCallCheck2.default)(this, TagTreeNode);
    (0, _defineProperty2.default)(this, "_init", void 0);
    (0, _defineProperty2.default)(this, "_ownedNodes", new Map());
    (0, _defineProperty2.default)(this, "_ownedByTag", new Map());
    this._init = init;

    this._init.executor({
      node: this,
      addOwnedNode: function addOwnedNode(tag, node) {
        _this._ownedNodes.set(node, tag);

        var entry = _this._ownedByTag.get(tag);

        if (!entry) {
          entry = _this._createTagEntry();

          _this._ownedByTag.set(tag, entry);
        }

        var _entry = entry,
            controller = _entry.controller;
        controller.add(node);
      },
      removeOwnedNode: function removeOwnedNode(tag, node) {
        _this._ownedNodes.delete(node);

        var entry = _this._ownedByTag.get(tag);

        if (!entry) throw new Error('tag not owned');
        var controller = entry.controller;
        controller.remove(node);
      },
      end: function end() {
        _this._ownedByTag.forEach(function (_ref) {
          var controller = _ref.controller;
          controller.end();
        });
      }
    });

    this._init.executor = function () {}; // release reference

  }

  (0, _createClass2.default)(TagTreeNode, [{
    key: "_createTagEntry",
    value: function _createTagEntry() {
      return _liveSet.default.active();
    }
  }, {
    key: "getValue",
    value: function getValue() {
      return this._init.value;
    }
  }, {
    key: "getParent",
    value: function getParent() {
      return this._init.parent || null;
    }
  }, {
    key: "getOwnedByTag",
    value: function getOwnedByTag(tag) {
      var entry = this._ownedByTag.get(tag);

      if (!entry) {
        if (!this._init.ownedTags.has(tag)) {
          throw new Error("tag not owned: ".concat(tag));
        }

        entry = this._createTagEntry();

        this._ownedByTag.set(tag, entry);
      }

      return entry.liveSet;
    }
  }, {
    key: "getOwned",
    value: function getOwned() {
      var m = new Map();

      this._ownedByTag.forEach(function (_ref2, tag) {
        var liveSet = _ref2.liveSet;
        m.set(tag, liveSet);
      });

      return m;
    }
  }, {
    key: "getTag",
    value: function getTag() {
      var parent = this._init.parent;
      return parent ? parent.getTagOfOwnedNode(this) : null;
    }
  }, {
    key: "ownsNode",
    value: function ownsNode(node) {
      return this._ownedNodes.has(node);
    }
  }, {
    key: "getTagOfOwnedNode",
    value: function getTagOfOwnedNode(node) {
      var tag = this._ownedNodes.get(node);

      if (tag == null) throw new Error('node not owned');
      return tag;
    }
  }]);
  return TagTreeNode;
}();

exports.default = TagTreeNode;
module.exports = exports.default;
module.exports.default = exports.default;
